influxdb3_cache/src/meta_cache/cache.rs

```diff
@@ -0,0 +1,513 @@
+                .collect(),
+        )
+    }
+
+    /// Prune nodes from within the cache
+    ///
+    /// This first prunes entries that are older than the `max_age` of the cache. If the cardinality
+    /// of the cache is still over its `max_cardinality`, it will do another pass to bring the cache
+    /// size down.
+    pub(crate) fn prune(&mut self) {
+        let before_time_ns = self.expired_time_ns();
+        let _ = self.data.remove_before(before_time_ns);
+        self.state.cardinality = self.data.cardinality();
+        if self.state.cardinality > self.max_cardinality {
+            let n_to_remove = self.state.cardinality - self.max_cardinality;
+            self.data.remove_n_oldest(n_to_remove);
+            self.state.cardinality = self.data.cardinality();
+        }
+    }
+
+    /// Get the nanosecond timestamp as an `i64`, before which, entries that have not been seen
+    /// since are considered expired.
+    fn expired_time_ns(&self) -> i64 {
+        self.time_provider
+            .now()
+            .checked_sub(self.max_age)
+            .expect("max age on cache should not cause an overflow")
+            .timestamp_nanos()
+    }
+
+    /// Get the arrow [`SchemaRef`] for this cache
+    pub(crate) fn arrow_schema(&self) -> SchemaRef {
+        Arc::clone(&self.schema)
+    }
+
+    /// Compare the configuration of a given cache, producing a helpful error message if they differ
+    pub(crate) fn compare_config(&self, other: &Self) -> Result<(), anyhow::Error> {
+        if self.max_cardinality != other.max_cardinality {
+            bail!(
+                "incompatible `max_cardinality`, expected {}, got {}",
+                self.max_cardinality,
+                other.max_cardinality
+            );
+        }
+        if self.max_age != other.max_age {
+            bail!(
+                "incompatible `max_age`, expected {}, got {}",
+                self.max_age.as_secs(),
+                other.max_age.as_secs()
+            )
+        }
+        if self.column_ids != other.column_ids {
+            bail!(
+                "incompatible column id selection, expected {}, got {}",
+                self.column_ids
+                    .iter()
+                    .map(|id| id.to_string())
+                    .collect::<Vec<_>>()
+                    .join(","),
+                other
+                    .column_ids
+                    .iter()
+                    .map(|id| id.to_string())
+                    .collect::<Vec<_>>()
+                    .join(","),
+            )
+        }
+
+        Ok(())
+    }
+}
+
+/// A node in the `data` tree of a [`MetaCache`]
+///
+/// Recursive struct holding a [`BTreeMap`] whose keys are the values nested under this node, and
+/// whose values hold the last seen time as an [`i64`] of each value, and an optional reference to
+/// the node in the next level of the tree.
+#[derive(Debug, Default)]
+struct Node(BTreeMap<Value, (i64, Option<Node>)>);
```

Comment:
Just for my clarity, does this mean each level in the tree will correspond to a column (in turn a `BTreeMap`)? If that's the case,
- does the filtering take advantage of the fact that it doesn't need to scan subsequent levels if the query doesn't include it, for eg. `select region from metadata_cache('cpu')` will not require `host` to be scanned. It sounds like something we'll do, but I couldn't figure out where we short circuit it.
- do we need to support out of band queries, like `select host from metadata_cache('cpu')` (note: there's no where clause for `region`)?

Reply:
> does the filtering take advantage of the fact that it doesn't need to scan subsequent levels if the query doesn't include it, for eg. select region from metadata_cache('cpu') will not require host to be scanned. It sounds like something we'll do, but I couldn't figure out where we short circuit it.

It does not. That would be "projection pushdown" which we could accomplish by handling the `projection` arg passed in the `TableProvider::scan` method. I did not attempt that yet. I will add an issue to tackle this since it is definitely doable but trickier than I want to handle on this PR and some requirements might need refining.

> do we need to support out of band queries, like select host from metadata_cache('cpu') (note: there's no where clause for region)?

It does handle this now, and [there is a test](https://github.com/influxdata/influxdb/blob/20d09a8dda5ac42a6cb388b15edc169f9fbbd709/influxdb3_cache/src/meta_cache/mod.rs#L681-L706) that does this, but it is doing a full scan of the cache hierarchy, and as I note in that test, then DataFusion does the filtering/projection - so this is not optimal.

We might be able to improve handling of this if we tackle projection pushdown, but will need to decide some things, e.g., if the output should be sorted or not.

Reply:
https://github.com/influxdata/influxdb/issues/25584